File: leetcode/0030/findSubstring.swift
func findSubstring(_ s: String, _ words: [String]) -> [Int] {
    if words.count == 0 {
        return []
    }
    let w_len = words[0].count, s_len = s.count
    let total_len = w_len * words.count // 子串的长度
    var wordDict = [String: Int]()
    for word in words {
        if let num = wordDict[word] {
            wordDict[word] = num + 1
        } else {
            wordDict[word] = 1
        }
    }
        
    var ans = [Int]()
    for offset in 0..<w_len {
        var leftOffset = offset
        let offsetMax = s_len - total_len
        while leftOffset <= offsetMax {
            var tempDict = wordDict
            var match = true
            
            var range = [Int]()
            var top = leftOffset + total_len
            while top > leftOffset {
                range.append(top)
                top -= w_len
            }
            
            var tempOffset = 0
            for hi in range {
                tempOffset = hi
                let word = s[hi-w_len, hi]
                if let num = tempDict[word] {
                    if num == 0 {
                        match = false
                        break
                    } else {
                        tempDict[word]! -= 1
                    }
                } else {
                    match = false
                    break
                }
            }
            if match {
                ans.append(leftOffset)
            }
            leftOffset = tempOffset
        }
    }
    
    return ans
}

extension String {
    subscript (left: Int, right: Int) -> String {
        get {
            let startIndex = self.index(self.startIndex, offsetBy: left)
            let endIndex = self.index(self.startIndex, offsetBy: right)
            return String(self[startIndex..<endIndex])
        }
    }
}


print(findSubstring("barfoothefoobarman", ["foo", "bar"]))

File: leetcode/0030/findSubstring.py
def findSubstring(s: str, words: [str]) -> [int]:
    if not words:return []
    w_len, s_len = len(words[0]), len(s)
    t_len = w_len * len(words)  # 子串的长度

    word_dict = {}  # words的哈希表
    for word in words:
        word_dict[word] = word_dict.get(word, 0) + 1

    ans = []
    for offset in range(w_len):
        lo, lo_max = offset, s_len - t_len
        while lo <= lo_max:
            tmp_dict = word_dict.copy()
            match = True
            # 从尾到头搜索单词
            for hi in range(lo + t_len, lo, -w_len): 
                # 切割出一个单词
                word = s[hi - w_len: hi]
                # 如果这个单词不在tmp_dict中
                if word not in tmp_dict or tmp_dict.get(word, 0) == 0:
                    match = False
                    break   # 当前单词不符合要求 直接停止这个子串的搜索
                tmp_dict[word] -= 1 # 如果在
            if match:
                ans.append(lo)
            lo = hi # 对lo直接赋值
    return ans

print(findSubstring("barfoothefoobarman", ["foo", "bar"]))
